From: SGI
Subject: lkcd: Fix overrun when num_mbanks > MAX_NUMNODES
References: 132221, 168907
Patch-mainline: 2.6.17

Fo large numa systems with more than one bank of memory/node, we sort of run
off the end of the array.  This is the situation that our customer was running
into.  This doesn't happen for systems that have a total number of dimms*nodes
that is less than MAX_NUMNODES, however for system that have 512 nodes, they
will run off the end of the array.

Acked-by: Chris Mason <mason@suse.com>
Acked-by: Raymund Will <rw@suse.de>

Index: linux/drivers/dump/dump_methods.h
===================================================================
--- linux.orig/drivers/dump/dump_methods.h	2005-11-15 16:07:46.999141079 -0800
+++ linux/drivers/dump/dump_methods.h	2005-11-15 22:45:22.423236481 -0800
@@ -47,6 +47,8 @@
 
 #define MAX_PASSES 	6
 #define MAX_DEVS	4
+#define MAX_MBANKS_PER_NODE	8
+#define MAX_MBANKS	(MAX_NUMNODES * MAX_MBANKS_PER_NODE)
 
 
 /* To customise selection of pages to be dumped in a given pass/group */
@@ -54,7 +56,7 @@
 	char name[32];
 	int (*selector)(int, unsigned long, unsigned long);
 	ulong level_mask; /* dump level(s) for which this filter applies */
-	loff_t start[MAX_NUMNODES], end[MAX_NUMNODES]; /* location range applicable */
+	loff_t start[MAX_MBANKS], end[MAX_MBANKS]; /* location range applicable */
 	ulong num_mbanks;  /* Number of memory banks. Greater than one for discontig memory (NUMA) */
 };
 
Index: linux/drivers/dump/dump_scheme.c
===================================================================
--- linux.orig/drivers/dump/dump_scheme.c	2005-11-15 16:07:46.855584793 -0800
+++ linux/drivers/dump/dump_scheme.c	2005-11-15 22:48:02.801970729 -0800
@@ -370,6 +370,11 @@
                         if(!(__dump_page_valid(loc >> PAGE_SHIFT)))
                                 continue;
 
+                        if (i == MAX_MBANKS) {
+				printk("Too many mbanks.  Dump will be incomplete.\n");
+				break;
+			}
+
                         /* We found a valid page. This is the start */
                         filter->start[i] = loc;
 
@@ -395,8 +400,8 @@
                 for(i = 0; i < dump_config.dumper->filter->num_mbanks; i++) {
                         filter->start[i] = dump_config.dumper->filter->start[i];
                         filter->end[i] = dump_config.dumper->filter->end[i];
-                        filter->num_mbanks = dump_config.dumper->filter->num_mbanks;
                 }
+                filter->num_mbanks = dump_config.dumper->filter->num_mbanks;
         }
 }
 #endif
